rust-runtime/aws-smithy-types/src/date_time/mod.rs

```diff
@@ -322,6 +323,120 @@ pub enum Format {
     EpochSeconds,
 }
 
+#[cfg(all(test, aws_sdk_unstable, feature = "serialize"))]
+mod ser {
+    use super::*;
+    use serde::ser::SerializeTuple;
+
+    impl serde::Serialize for DateTime {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            if serializer.is_human_readable() {
+                match self.fmt(Format::DateTime) {
+                    Ok(val) => serializer.serialize_str(&val),
+                    Err(e) => Err(serde::ser::Error::custom(e)),
+                }
+            } else {
+                let mut tup_ser = serializer.serialize_tuple(2)?;
+                tup_ser.serialize_element(&self.seconds)?;
+                tup_ser.serialize_element(&self.subsecond_nanos)?;
+                tup_ser.end()
+            }
+        }
+    }
+}
+
+#[cfg(all(test, aws_sdk_unstable, feature = "deserialize"))]
+mod der {
+    use super::*;
+    use serde::de::Visitor;
+    use serde::Deserialize;
+    struct DateTimeVisitor;
+
+    enum VisitorState {
+        First,
+        NotFirst,
+    }
+
+    struct NonHumanReadableDateTimeVisitor {
+        state: VisitorState,
+        seconds: i64,
+        subsecond_nanos: u32,
+    }
+
+    impl<'de> Visitor<'de> for DateTimeVisitor {
+        type Value = DateTime;
+        fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+            formatter.write_str("expected RFC-3339 Date Time")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+            E: serde::de::Error,
+        {
+            match DateTime::from_str(v, Format::DateTime) {
+                Ok(e) => Ok(e),
+                Err(e) => Err(serde::de::Error::custom(e)),
+            }
+        }
+    }
+
+    impl<'de> Visitor<'de> for NonHumanReadableDateTimeVisitor {
+        type Value = Self;
+        fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+            formatter.write_str("expected (i64, u32)")
+        }
+
+        fn visit_i64<E>(mut self, v: i64) -> Result<Self::Value, E>
+        where
+            E: serde::de::Error,
+        {
+            match self.state {
+                VisitorState::First => {
+                    self.seconds = v;
+                    self.state = VisitorState::NotFirst;
+                }
+                _ => return Err(serde::de::Error::custom("First value must be i64")),
+            };
+            Ok(self)
+        }
+        fn visit_u32<E>(mut self, v: u32) -> Result<Self::Value, E>
+        where
+            E: serde::de::Error,
+        {
+            match self.state {
+                VisitorState::NotFirst => self.subsecond_nanos = v,
```

Comment:
May be good to add a third state that panics if it's encountered as a sanity check.

Reply:
this one's fixed